wallet-new/src/Cardano/Wallet/Kernel/Actions.hs

```diff
@@ -147,13 +150,51 @@ initialWorkerState = WalletWorkerState
     , _lengthPendingBlocks = 0
     }
 
--- | Start up a wallet worker; the worker will respond to actions issued over the
---   returned channel.
-forkWalletWorker :: WalletActionInterp IO b -> IO (WalletAction b -> IO ())
-forkWalletWorker ops = do
-    c <- newChan
-    link =<< async (walletWorker c ops)
-    return (writeChan c)
+
+-- | Start a wallet worker in backround who will react to input provided via the
+-- 'STM' function, in FIFO order.
+--
+-- After the given continuation returns (successfully or due to some exception),
+-- the worker will continue processing any pending input before returning,
+-- re-throwing the continuation's exception if any.
+--
+-- Usage of the obtained 'STM' action after the given continuation has returned
+-- will fail with an exception.
+withWalletWorker
+  :: (MonadIO m, Ex.MonadMask m)
+  => WalletActionInterp IO a
+  -> ((WalletAction a -> STM ()) -> m b)
+  -> m b
+withWalletWorker wai k = do
+  -- 'mDone' is full if the worker finished.
+  mDone :: MVar (Either Ex.SomeException ()) <- liftIO newEmptyMVar
```

Comment:
This is basically exactly what `waitCatch` gives you from the `async` package, at least seems like so, superficially. Why do you feel this MVar-based solution is better?